### ISSUE TITLE
New PR CI Architecture to support more runners

### ISSUE THREAD

**Issue by felipevolpone** [freeipa/freeipa-pr-ci#159]
The PR CI as it is now, was not designed to support lots of runners (60+ runners), with that in mind, I would like to propose a new architecture for the project.

![new pr ci arch 1](https://user-images.githubusercontent.com/1590527/36326361-f7d337c8-1341-11e8-9850-a6490bb8ed8e.png)

This is the workflow described in details:

1. GitHub (GH) trigger the web server when a PR is created or when a PR it's modified (using webhooks).
2. With the PR information, the web server will read the `.freeipa-pr-ci.yaml` spec file and will create a Task to each test described on it. A Task is basically an object with info of the test, the priority, if that Task depends on another one, etc.
   1. An important detail here. When a PR is created, we first need to run the build process, then the other tests can run. At the first time, the web server would read the PR info, see that it doesn't have any tasks created (in its task list), and it would create only the build Task and submit it to the RabbitMQ.
When the build process is done, the Publisher will update GH saying that the build succeeded. With [this GH webhook](https://developer.github.com/v3/activity/events/types/#statusevent) (it's triggered when a status of a commit changes), the web server can check that the build is done and then create the other Tasks.

3. The web server will add the Tasks in the `TASKS_QUEUE` on the RabbitMQ. This queue is only for tasks that the runners will consume and run. Important to say that we can have different kinds of tests to run, some of then consume less resource, but others demand bigger runners (VMs with more RAM and vCPU). So, we will need different queues for each kind of tests to run. Tests that consume less resource, will be added to the "medium tasks queue" and the bigger ones to "big tasks queue". In practice, we will not have only the  `BITASKS_QUEUE`, but  `BIG_TASKS_QUEUE`,  `SMALL_TASKS_QUEUE` and so on.
4. The runners consume the `TASKS_QUEUE` and each runner will get a task to run. Important to say that RabbitMQ has its own mechanisms to prevent that a message in a queue could be consumed by two consumers. With the Task on hands, the runner knows what do to. 
5. After running the tests, the runner will create a Result object, that will be added to the `RESULTS_QUEUE` on the RabbitMQ.
6. A Publisher (this is could be a service running on some machine) that consumes the `RESULTS_QUEUE` queue will get the Result object and publish the logs on `FedoraPeople` and change the GH task list, showing if the tests succeed or not (as we already do today)

Advantages:
 - This way we will never hit the API rate limit; given that we will use webhooks
 - It's totally uncoupled from GH. In the new arch. the tasks will be created by the Web Server based on the info received by GH, however, if sometime in the future we move from GH, we would just need to change the code that talks with the PR list.

Disadvantages:
 - It's more complex and it has more point of failures
 - It's harder to deploy than it is today. Today we just need to deploy the runners. With this new proposal, we would need to deploy: the RabbitMQ, the web server, the runners and the publisher. All this can be automated.
 - It's harder to debug, given that it has more pieces on the game.

Challenges:
* Serialization of the object (Task and Result). It's not obvious to serialize and deserialize big objects like classes and so on. So a default "protocol" could/should be used, like only transfer JSON files.

PS:
* The items II, III, IV and VI are in blue color to show that they belong to our infrastructure.
* The items II and IV could be deployed on the same machine.
* The item IV could run as a service

**Comment by tomaskrizek** [comment]
Like you mentioned, this significantly affects the complexity of deploying your own runners. I know there are some people (@celestian) who use the PR CI in QE to run their tests. Perhaps it would be a good idea to branch / tag the current architecture so they're still able to deploy & use it in the same way (as long as it continues to work...)

**Comment by celestian** [comment]
So, this new architecture adds new (central) units to deployment. It could be nice to have our current workflow still available. The former reason of PR CI was to support our community with simply/easy-to-use PR CI.

If I understand correctly, current project is about:
1) how to orchestrate runner;
2) github PR logic.

In my point of view it could be possible to separate those two steps. And reuse step 1 for both, old and new architecture. I mean it should be possible to setup CI as old one or as new one. This could be really nice.

@felipevolpone do you think that it is possible, please?

**Comment by slaykovsky** [comment]
Hello @felipevolpone  :)

Your CI architecture looks very good, but I think it's overkill.

I think we can have only one endpoint reachable from the outside world so we can host it even on free OpenShift tier. I really like points 1 and 2 in your workflow but also I think it's good to create initial GitHub API polling there on application startup. Another thing I want to propose that every runner should have a persistent connection (such as WebSockets) or polling to a controller so we can have stats from every runner and monitor the whole infrastructure. I don't want to start any implementation wars just yet as we need to have a TLD.

I made an ugly picture of how I see the design of CI:

![123](https://user-images.githubusercontent.com/1875374/36384047-4246d000-158e-11e8-8b5b-d9b049e77620.png)

So let me explain the workflow here, it's very similar to yours :)

1. When the controller app is started we need to build up a tasks queue. As we didn't have any events there yet, we need to poll the GitHub API.
2. Every runner should connect to a controller and give its stats (free CPU, ram, running tasks, etc) there.
3. If it any new tasks in a queue, the controller should distribute them across the runners.
4. When the payload from GitHub will come (new PR, changed PR, closed PR, rerun label, build success, etc) controller will deal with it and, as it was described above, distribute tasks though runners or send a message to kill the running ones. So it's well described in your comment in 2.i that for a new or changed PR we need to init a build task and then wait for its status.
5. We can use publisher service you described for uploading artefacts to some storage, but also we can leave this as a runner's job.
6. As a runner finished the task it should notify the controller and then the controller will update the status on GitHub.

So yeah, that's all I want to say right now.

Thanks.

**Comment by tomaskrizek** [comment]
There is a clear conflict between usability of this project either as the main FreeIPA's CI or as a local/community testing tool. FreeIPA CI requires scalability, because it can have tons of runners. Therefore API limit becomes an issue, thus the new architecture.

For local or small scale testing for QE/community/developers, the API limit is not at issue at all and the primary concern is how simple it is to use. The current design favors this use case, as the runner deployment requires just a single hypervisor, and the system doesn't require any extra configuration, just your github token.

Even if the new design would allow deployment of all the key components on a single machine, extra configuration of the web hooks would be required. This is also an issue when your machine is not reachable from the internet, because you don't have anything to point the webhook to. With the new architecture, the local/small scale testing use case will be very hard to deploy and maintain.

---

@slaykovsky It's not clear whether you'd want to use web socket for GitHub API or to poll it. Either case could work, but polling the API gets rids of the problem described above (deploying local runners). 

I prefer the grouping of the runners in a single controller like you suggest, because it still makes it simpler to deploy and maintain, while still addressing the scalability issue. Polling  the GitHub PR queue with a single (or few) controllers shouldn't hit the API limit and it would still allow deployment of local runners. (Especially if there was an option to deploy controller & runner on the same machine).

**Comment by davidkupka** [comment]
I understand the issue with requests limitation but from my point of view this proposal doesn't fit the requirements. IIRC there was requirements of no single point of failure, no publicly accessible component and multiple independent* deployments cooperating on single repository. Were these requirements removed?

\* in different networks/managed by different people

**Comment by stlaz** [comment]
@davidkupka There's a need to react to the bottleneck created by GitHub API limits which are often hit as the API was never really meant to be used for CI.
On the other hand, it might be possible (and I would prefer that) to implement the controller from @slaykovsky's proposal as a separate project and then run PR CI runners against it, thus keeping the original design aims you mention (the requirements were a bit different).

**Comment by felipevolpone** [comment]
With all said, I think we should go with @slaykovsky proposal, but still pooling the API (in order to still be able to have runners without outside-world-contact). I'm just confused by WebSockets part, how will you control which runner should run a particular test? Given that we have different types of runner and tests.

**Comment by felipevolpone** [comment]
As it was also said, my proposal is way more complex than the other proposal. I just would like to share what I had in mind when designing it:

* We would get rid of GitHub coupling: with that, the change to have something more close of Fedora CI Standard Test Interface would be easier.
* The RabbitMQ Queue could be an interface to a CI runner in general. If we want it, we could create some kind of GUI to create Tasks there, because as I already pointed, it would not depend on creating PRs on GH. As an example, imagine something like we already have for Jenkins. I thought about that because it would also decrease the conflict between Q&A and dev people about PR CI.

Does it worth to have such complex architecture in order to have the advantages above? I don't know. I think it's up to the team decide.

**Comment by pvoborni** [comment]
In a car, on the way home, I was thinking whether there is a way to avoid changing architecture to keep the original properties while supported o lot of open PRs and a lot of runners without hitting the limit.

So I came up with following solution, please read and say what aspect I did forget ;). Otherwise it is simpler then the thing which we have now (no special requirement for caching).

```python
"""
Use cases of PR-CI github comunicator:

1. Get next task to run.
2. Upload task result.
3. Get next newly open PR and generate tasks.

* rate limit is 5000 requests per hour <https://developer.github.com/v3/#rate-limiting>
* get all open prs returns list of prs including labels <https://developer.github.com/v3/pulls/#list-pull-requests>
* tasks are controlled via status API https://developer.github.com/v3/repos/statuses/
"""

# github labels used by the solution (can be reworded)
CI_PENDING = "ci_pending" # allows to ignore old, already tested PRs by runners
CI_DONE = "ci_done"  # allows to ignore old, already tested PRs  by runners
CI_NOT_ON_WHITELIST = "ci_needs_approval" # notification for core team
CI_RERUN = "re-run"

not_for_me = [] # prs this runner should ignore (in memory cache) - support of various sizes of runners

def get_prs_with_tasks():
    prs = github.get_prs("open") # 1 API call, can control cache with etag (optimization)

    # get only prs with task to run
    prs_to_test = [pr for pr in prs if CI_PENDING not in pr.lables]
    return prs_to_test


# in total 2 API calls per task
def get_next_task():
    prs = get_prs_with_tasks()
    prs = [pr for pr in prs if pr not in not_for_me]
    for pr in prs
        statuses = github.get_statuses(pr._links.statuses) # 1 API call
        status = None
        while not status and statuses
            status = get_random_status()
            if i_am_able_to_handle_this_task(status)
                return status # we got status(task) to run, great
            else
                statuses.remove(status)
                status = None
        not_for_me.append(pr) # next run runner won't even try to get statuses for it
    return None # nothing to test


# 2-3 API calls
def run_task(status):
    result  = do_magic(status) # run the task
    github.status.update(result) # 1 API CALL
    statuses =  github.get_statuses(status.ref) # 1 API CALL
    if all_done(status):
        github.pr_update(add_label=CI_DONE, remove_label=CI_PENDING) # 1 API CALL??


# 1..n (number of task) API calls
def generate_tasks(pr):
    tasks = get_tasks(pr)
    generated = False
   
    for tasks in tasks
       github.status.add(task)
       generated = True

    if CI_RERUN in pr.labels:
        github.pr_update(remove_label=CI_PENDING)

    if generated:
        github.pr_update(add_label=CI_PENDING)
    else:
         github.pr_update(add_label=CI_DONE)

    return generated

# 2..n (number of task) API calls
def generate_tasks_for_new_pr():
    prs = github.get_prs("open") # 1 API call

    # PRs which requires task generation don't have any PR-CI related label
    prs_to_generate =  [pr for pr in prs if CI_PENDING not in pr.labels and 
                                            CI_DONE not in pr.labels]
    generated = False

    pr = get_random(prs_to_generate)
    # Use case: PR author is core contributor (on whitelist)
    if pr.author in white_list:
        generated = generate_tasks(pr) # 1-n (number of task) API calls
    # Use case: author is not on whitelist but core contributor allowed it
    else if CI_RERUN in pr.labels and pr.labels[CI_RERUN].author in white_list:
        generated = generate_tasks(pr) # 1-n (number of task) API calls
    # Use case: author is not on whitelist and we need to let know core contributers that
    # there is action pending
    else: 
        github.pr_update(add_label=CI_NOT_ON_WHITELIST)

    return generated


def main():
    while True:
        # prefer to run task - older prs will be tested sooner
        task = get_next_task() # 2 API calls
        if task:
            run_task(task)# 2-3 API calls
            # there is no wait, so runner get pickup next task immediately
            continue 

        # if there are not tasks, generate new
        generated = generate_tasks_for_new_pr() # 2..n (number of task) API calls

        # if nothing was generated then there is no new PR, so wait for opening a new PR or
        # re-run label this wait should be based on number of runners.
        # If there is a lot of runners thena  lot of runners will try to generate new tasks
        # therefore this wait can be longer. E.g 5s x NUMBER_OF_RUNNERS. Given that runners
        # don't know this info and there is not way how to get it then admin needs to cofigure
        # it during runner spawning
        # Default value: 15s - let's assume that for personal usage there will be only 1-2 
        # runners waitinig 7s-20s is OK.
        if not generated:
            wait(randomize(GENERATION_WAIT))
        # if generated, then don't wait and get task immediately

"""
Properties of this algoritm:

If there are tasks to run then it will consume only 4 API calls. Ie. with other overhead
of 1000 API request(reserve for future) we are able to handle 1000 ((5000-1000)/4) tasks 
in 1h. This would require infra with approx 3000 vCPUs and 6000GB RAM -> we are fine 
for quite some time.

If there is nothing to run then there will be approx 1 API call per 7s, so 514 per hour. 
This number is more or less constant provided GENERATION_WAIT is well set.

Generation of tasks is N API calls, but this will be the same with all solutions - something
needs to generate them. API limit won't be reach if 800 tasks is generated for 1h. But his
would, again, require wery big infra (approx 500+ runners).

Summary: it should be able to handle unlimed number of open PRs and unlimed number of runners.
The only limitaion is number of newly open PRs per hour and number of tasks there. But this
limitation is present also in any "central controller" solution.
"""

```

**Comment by slaykovsky** [comment]
@felipevolpone @pvoborni  we can get all the info we need as one HTTP GET to GraphQL GitHub API endpoint, something like this:


```graphql
{
  repository(owner: "freeipa", name: "freeipa") {
    pullRequests(last: 100, states: OPEN) {
      nodes {
        number
        author {
          login
        }
        labels(last: 5) {
          nodes {
            name
          }
        }
        commits(last: 1) {
          nodes {
            commit {
              oid
              status {
                contexts {
                  context
                  description
                  state
                  targetUrl
                }
              }
            }
          }
        }
      }
    }
  }
}
```

**Comment by pvoborni** [comment]
Yes, the GraphQL could be an optimization for the algorithm above. Or maybe also the current way. But with current way, I'm not sure (don't know it that well).

**Comment by felipevolpone** [comment]
@pvoborni I like your proposal and it could be away to avoid rewritting the arch. I only miss one aspect there:
* We need to update the status in GH to control that another runner should not get that task. To do this, we could add the line below to the `get_next_task` method:

```diff
def get_next_task():
    prs = get_prs_with_tasks()
    prs = [pr for pr in prs if pr not in not_for_me]
    for pr in prs
        statuses = github.get_statuses(pr._links.statuses) # 1 API call
        status = None
        while not status and statuses
            status = get_random_status()
            if i_am_able_to_handle_this_task(status)
+              github.update_status(task) # add this line
                return status # we got status(task) to run, great
```

@pvoborni @slaykovsky About the GraphQL change: I'm pro for that, but I think we would be rewriting too much code in the project (if we also will rewrite the arch). Also, it would not add any value *now*, because the bottleneck is not in the 5k req/hour as I already mentioned before. The bottleneck is in the 1500 cached reqs per minute, that we do to check if the PR changed.

**Comment by pvoborni** [comment]
@felipevolpone yes, ` github.update_status(task)` would be required. I missed to write it. Thanks. When I talked to @slaykovsky I realized that my proposal would require a similar amount of work as a similar solution which uses GraphQL. So I would go for rewriting the github communication classes to use GraphQL while using similar logic. The advantage of GraphML usage is that we don't need to use `CI_PENDING` and `CI_DONE` labels as they were only a way how to limit number of requests.

As for general usage or running other tasks not involving GH I'll post some ideas soon. But it should work well together with direct GitHub usage solution.
